amaranth/lib/fifo.py

```diff
@@ -477,24 +478,26 @@ def elaborate(self, platform):
         # are reset to 0 violate their Gray code invariant. One way to handle this is to ensure
         # that both sides of the FIFO are asynchronously reset by the same signal. We adopt a
         # slight variation on this approach - reset control rests entirely with the write domain.
-        # The write domain's reset signal is used to asynchronously reset the read domain's
-        # counters and force the FIFO to be empty when the write domain's reset is asserted.
-        # This requires the two read domain counters to be marked as "reset_less", as they are
-        # reset through another mechanism. See https://github.com/amaranth-lang/amaranth/issues/181
-        # for the full discussion.
+        # The write domain reset signal is used to asynchronously force the FIFO to be empty. The
+        # write domain counters are overwritten with the value of the read domain counters. This
+        # requires the two write domain counters to be marked as "reset_less", as they are reset
+        # through another mechanism.
+        # See https://github.com/amaranth-lang/amaranth/issues/181 for the full discussion.
         w_rst = ResetSignal(domain=self._w_domain, allow_reset_less=True)
         r_rst = Signal()
 
         # Async-set-sync-release synchronizer avoids CDC hazards
         rst_cdc = m.submodules.rst_cdc = \
             AsyncFFSynchronizer(w_rst, r_rst, o_domain=self._r_domain)
 
-        # Decode Gray code counter synchronized from write domain to overwrite binary
-        # counter in read domain.
+        # Decode Gray code counter synchronized from read domain to overwrite binary counter in
+        # write domain.
+        with m.If(w_rst):
+            m.d[self._w_domain] += produce_w_gry.eq(consume_w_gry)
+            m.d[self._w_domain] += produce_w_bin.eq(_gray_decode(consume_w_gry))
```

Comment:
This does not work: if a write domain reset happens one clock cycle after a read, then the next value of `consume_w_gry` is still inside an `FFSynchronizer`.